### ISSUE TITLE
mat-icon shows text before showing properly

### ISSUE THREAD

**Issue by Springrbua** [angular/material2#12171]
#### Bug, feature request, or proposal:
Bug or feature request.

#### What is the expected behavior?
When using `<mat-icon>home</mat-icon>`, nothing should be displayed until the icon is loaded properly. As soon as the icon is loaded, it should render.

#### What is the current behavior?
Depending on the speed of your connection, it is possible, that you see the text (for example `home`) instead of the icon.

#### What are the steps to reproduce?

1. Start the Angular Material Icon example in [Stackblitz](https://stackblitz.com/angular/yogqmggjokx).
It might be better to open it in it's own tab.
2. Open the Developer-Tools (in Chrome) and under the `Network`-Tab choose a very slow network throttling profile. This is needed, as the sample app is very small and therefore the problem is not noticeable with normal connections.
3. Reload the page and wait until it is loaded. You will see something like this:
![image](https://user-images.githubusercontent.com/18258317/42627235-b224a5e2-85cc-11e8-89d9-29b94933423a.png)
After some time it shows the icon:
![image](https://user-images.githubusercontent.com/18258317/42632653-2e332d8c-85de-11e8-95b2-5a18c09c5fcb.png)

#### What is the use-case or motivation for changing an existing behavior?
In an app I am currently working on, this happens pretty often. Even if I only change the route (which does not trigger a reload of the page) I often see the text of the icons before the icon is displayed properly, so it does not seem to be related to the loading of the icon font.

#### Which versions of Angular, Material, OS, TypeScript, browsers are affected?
I am using @angular/material 6.3.1 together with TypeScript 2.7.2 on a Windows 10 PC.  
The problem occures at least on Chrome, Opera, Firefox and Edge.

**Comment by Springrbua** [comment]
I just noticed, that the problem in my app was actually the change-detection. I am using `CdkPortal` and `CdkPortalOutlet` to dynamicaly change the ocntent of my app-bar (including some `mat-icon`s).  
As I am using `ChangeDetectionStrategy.OnPush`, I had to call `detectChanges()` after changing the content.
However, you can still see the text, while the font is loading. On normal connections it isn't really noticeable, but if you have a slow connection, the text stays for a while.
I guess it would be good, if the icons don't show anything as long as the used font is not fully loaded.
Not sure, if thats even possible...

**Comment by jelbourn** [comment]
Going to close this as "working as expected". The component on its own doesn't have a good way of knowing when the icon font is loaded. If your app knows when the assets are loaded, you can hide the icons until then, or you can potentially fall back to pngs or svgs.

**Comment by thw0rted** [comment]
@jelbourn would you reopen this if the component does in fact have a good way of [knowing when the font is loaded](https://developer.mozilla.org/en-US/docs/Web/API/FontFaceSet/load)?

    if ("fonts" in document && !document.fonts.check(iconFontName)){
      document.body.classList.add("mat-icon-font-missing");
      document.fonts.load(iconFontName)
        .then(() => document.body.classList.remove("mat-icon-font-missing"));
    }

CSS should be easy enough once you have that.  Of course, this doesn't avoid the FOUT for IE/Edge users, but [there's just no helping some people](https://wpdev.uservoice.com/forums/257854-microsoft-edge-developer/suggestions/6509785-css-font-loading).

**Comment by thw0rted** [comment]
Is the icon registry responsible for the default behavior today?  The docs say "By default, <mat-icon> expects the Material icons font" with a link to instructions on including the `@font-face` definition.  I was going to suggest a universal / top-level "icon font not loaded" class, but of course with the registry you can have an arbitrary number of `fontSet` so you'd need to do the above logic for each font in use.

And I see what you were saying in your original response, none of the current registry methods seem to care about the font name. Moreover, it appears to be possible to use font-based `mat-icon` without touching the registry at all -- I haven't done it myself, but it sounds like you just need to set the `fontSet` / `fontIcon` params.  I guess this sort of behavior would need to either be added to a current call (like an extra parameter for `registerFontClassAlias`) or just be exposed as a new method.

**Comment by jelbourn** [comment]
Yeah, the registry has the default icon font css class. So in addition to that class, it would need to know about the font name.

**Comment by anymos** [comment]
Hi there,

 document.body.classList.add("mat-icon-font-missing");
  document.fonts.load(iconFontName)

Please keep in mind that it might break universal if you refer to classList

**Comment by thw0rted** [comment]
When you say "break universal", I guess technically Material2 could support IE9 but that's the only outlier I see on [the list](https://caniuse.com/#feat=classlist) and frankly, if you're stuck on IE9 you have bigger problems in your life 😒

**Comment by kwalski** [comment]
just checking if there is any PR WIP on this?

**Comment by NodeSwapCeo** [comment]
> When you say "break universal", I guess technically Material2 could support IE9 but that's the only outlier I see on [the list](https://caniuse.com/#feat=classlist) and frankly, if you're stuck on IE9 you have bigger problems in your life 😒

Not sure about IE9, universal is server side rendering ;-) fortunately I am not stuck with IE9 aahahah

**Comment by benneq** [comment]
I think I got the same issue: `mat-icon` displays only text when using `OnPush` in the parent component. When `OnPush` is removed everything works.

Here's a Stackblitz example with angular 8 an angular material 8: https://stackblitz.com/edit/components-issue-ghjcpf?file=app/app.component.ts

**Explanation of the code:**
I have a custom `IfDirective` (which I have modified to show the problem). The directive changes every 2 seconds its `ViewContainer` contents.

**What you should be able to see:** 
The first 2 seconds you can see the `notifications` **icon**. After those 2 seconds the directive will clear the view container and displays the other template. This other template also contains a `mat-icon`, but it will always display the string instead of the icon.

When you remove `changeDetection: ChangeDetectionStrategy.OnPush` from the `AppComponent` everything works as expected. But that's no solution.

When you change the directive input variable `show` from within the `AppComponent` everything is working fine including `OnPush`, because in this case the component automatically runs the change detection, I guess.

The `mat-raised-button` with text behave normal. Animations, text, styles are in place. The problem only happens with `mat-icon`.

**Comment by mrmashal** [comment]
Is there a requirement for the font ligature identifier to be provided as the text inside the `<mat-icon>` tag? I think you can introduce an attribute for this function, just like the `svgIcon` or `fontIcon` attributes, and the html-rendering issue would be solved. e.g.:

`<mat-icon ligature="home"></mat-icon>`

**Comment by cbejensen** [comment]
Here's how I am solving it:

**Comment by cbejensen** [comment]
Here's how I am solving it, with [`fontfaceobserver`](https://github.com/bramstein/fontfaceobserver/blob/master/src/observer.js):

### app.component.ts
```typescript
export class AppComponent {
  constructor(renderer: Renderer2) {
    const materialIcons = new FontFaceObserver('Material Icons');
    materialIcons.load(null, 10000).then(() => renderer.addClass(document.body, 'material-icons-loaded'));
  }
}
```

### main.scss
```scss
// don't target mat-icon elements with a different font set
.mat-icon:not([fontSet]) {
  display: none;
  .material-icons-loaded & {
    display: inline-block;
  }
}
```

**Comment by thekarlbrown** [comment]
I am working with an Angular 8 project where I am utilizing mat-icon's that were experiencing FOUT (Flash of Unstyled Text) resulting in the icon text to be displayed for a brief moment while waiting for the Material Icon font families to load. My solution was to add the npm package [webfontloader](https://www.npmjs.com/package/webfontloader)

`npm install --save webfontloader`

Inside the app.module.ts:

`import * as WebFont from 'webfontloader';

WebFont.load({
  custom: { families: ['Material Icons', 'Material Icons Outline'], }
});`

And in the css/sass file where the font families are imported:

`.wf-loading body {
  display: none;
}`

**Comment by NoelLH** [comment]
Thanks for the discussion all, and suggested workarounds @cbejensen + @thekarlbrown.

I was seeing this issue in an Angular Universal app, I think most visibly on even faster connections during the handover from server to client JS (possibly a side effect / deeper issue around where things load from, but this was at least the most visible symptom).

I had to adapt these solutions to get server renders working. I tried the Typekit `webfontloader` first but found it had a `window` dependency, at least when loaded with `import`, which broke the SSR build.

Using `fontfaceobserver` got round this problem, but giving `Renderer2` a `document.body` ref led to a runtime crash on the server side, for (in retrospect) obvious reasons.

So my approach is now based on @cbejensen's but hiding the app component instead of the body, and using `ElementRef`:

`app.component.ts`:
```typescript
import { ElementRef, Inject, OnInit, Renderer2 } from '@angular/core';
// ...
export class AppComponent implements OnInit {
  constructor(
    private elementRef: ElementRef,
    @Inject(PLATFORM_ID) private platformId: Object,
    private renderer: Renderer2,
    // ...
  ) {}

  ngOnInit() {
    if (isPlatformBrowser(this.platformId)) {
      // Avoid flash of icon placeholder text during SSR -> client JS handover.
      const materialIcons = new FontFaceObserver('Material Icons');
      materialIcons.load(null, 10000)
        .then(() => this.renderer.addClass(this.elementRef.nativeElement, 'material-icons-loaded'));

      // ...existing client-side-only inits.
    } else {
      // On the server side, add the class to show icons immediately.
      this.renderer.addClass(this.elementRef.nativeElement, 'material-icons-loaded');
    }
  }
}
...
```

`styles.scss`:
```scss
...
app-root:not(.material-icons-loaded) {
  display: none;
}
...
```

Hope this helps anyone else trying to sort this with a Universal / dynamic SSR build.

**Comment by makitsune** [comment]
There's a really straightforward fix for this.

If you self-host your material icons (which I recommend doing with all your fonts)
https://google.github.io/material-design-icons # Setup Method 2. Self hosting

You'll find 2 css snippets of `@font-face` and `.material-icons`
In the `@font-face` css add `font-display: block` like this
```css
@font-face {
  font-family: 'Material Icons';
  font-style: normal;
  font-weight: 400;
  font-display: block;
  ...
```

This will stop the font from showing anything until it loads.

https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display

**Comment by mohamed-tebani** [comment]
> There's a really straightforward fix for this.
> 
> If you self-host your material icons (which I recommend doing with all your fonts)
> https://google.github.io/material-design-icons # Setup Method 2. Self hosting
> 
> You'll find 2 css snippets of `@font-face` and `.material-icons`
> In the `@font-face` css add `font-display: block` like this
> 
> ```css
> @font-face {
>   font-family: 'Material Icons';
>   font-style: normal;
>   font-weight: 400;
>   font-display: block;
>   ...
> ```
> 
> This will stop the font from showing anything until it loads.
> 
> https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display

Work for me !

**Comment by kfir124** [comment]
@mohamed-tebani fix was great for me, 1 line of css and just works! thanks

**Comment by ievgennaida** [comment]
Would be nice to allow hide text in case of failure.

**Comment by PhanindraNath** [comment]
> There's a really straightforward fix for this.
> 
> If you self-host your material icons (which I recommend doing with all your fonts)
> https://google.github.io/material-design-icons # Setup Method 2. Self hosting
> 
> You'll find 2 css snippets of `@font-face` and `.material-icons`
> In the `@font-face` css add `font-display: block` like this
> 
> ```css
> @font-face {
>   font-family: 'Material Icons';
>   font-style: normal;
>   font-weight: 400;
>   font-display: block;
>   ...
> ```
> 
> This will stop the font from showing anything until it loads.
> 
> https://developer.mozilla.org/en-US/docs/Web/CSS/@font-face/font-display

Update on two things

1. Google fonts support the `font-display` property in the URL itself. So, self-hosting is not mandatory.

> https://fonts.googleapis.com/icon?family=Material+Icons|Material+Icons+Outlined|Material+Icons+Round&display=block"

2. `font-display: block` gives only a short period of `block` after that it works like `swap`. which means, if you can throttle the network to `slow 3G`, you still see FOUT after some time.

**Comment by Huiet** [comment]
I know this isn't a great solution (since it's by no means reactive to when the actual icon is available for use), but it worked for our use case.

tl;dr: use ngclass and switch a boolean to display the icon after a certain time via setTimeout

1. define a class to hide the icon, in my case this was 
`.hide-icon {
  color: transparent;
}`

2. ng-class based on a boolean property you define
` <mat-icon [ngClass]="{'hide-icon': !initialized}">
{{hidePassword ? 'visibility_off' : 'visibility'}}
</mat-icon>`
          
 3. timeout to the rescue!
 set the variable as false initially, and change it to true after some amount of time (we use 2 seconds)
 ` ngAfterViewInit() {
                setTimeout(() => { this.initialized = true; this.cdr.detectChanges(); }, 2000);
  }`

**Comment by dillyboy** [comment]
Here is how I managed to minimize this issue. But note that it only works for Chrome and Edge, otherwise it will fallback to the usual behaviour. By using document.fonts.onloadingdone I check if the font I want is loaded. If it is loaded I create a class which will turn the opacity of the element 'mat-icon' to 1 which was initially set as 0 in the css.

On a css file I write this rule (which will get overridden by the class which will be defined in the script eventually)
`mat-icon { opacity: 0 };`

I added a script tag in my index html page.



    // This executes on chrome and EDGE only as tested
    document.fonts.onloadingdone = function (fontFaceSetEvent) {
      // console.log(fontFaceSetEvent.fontfaces) to see the available fonts which have been loaded
      // and change the font family name according to your font family requirement
      const fontName = 'Material Icons';
      if (fontFaceSetEvent.fontfaces.filter(i => i.family === fontName).length > 0) {
          addMakeIconsVisibleClass();
      }
    };

    // Fallback - call below function if not chrome (or EDGE)
    if (navigator.userAgent.toLowerCase().indexOf('chrome') === -1) {
      addMakeIconsVisibleClass();
    }

    function addMakeIconsVisibleClass() {
      let style = document.createElement('style');
      style.innerHTML = 'mat-icon { opacity: 1 !important }';
      document.getElementsByTagName('head')[0].appendChild(style);
    }

**Comment by markiewiczart** [comment]
As a very simple workaround you can preload the font via <link> in your index.html. For instance:

```
<head>
  <link rel="preload" href="assets/fonts/MaterialIcons/MaterialIcons-Regular.woff2" as="font" type="font/woff2">
</head>
```

**Comment by vangorra** [comment]
I'm using the following inside the AppComponent and hiding all content until the fonts have loaded. Seems to work well.
```
document.fonts.ready.then(() => (this.isFontsLoaded = true));
```

**Comment by tomaszs** [comment]
I came here after searching for a way to hide the text on a dynamically opened component that contains mat-tab.
The list of possible hacks here is nice.

But if it comes to the main issue: when will it be fixed?

**Comment by thiagosbrito** [comment]
For me the best solutions was to add `&display=block` at the end of import file
`@import url('https://fonts.googleapis.com/css2?family=Material+Symbols+Outlined:opsz,wght,FILL,GRAD@20..48,100..700,0..1,-50..200&display=block');`

**Comment by dhdaines** [comment]
> There's a really straightforward fix for this.
> 
> If you self-host your material icons (which I recommend doing with all your fonts) https://google.github.io/material-design-icons # Setup Method 2. Self hosting

Note that these instructions are *wrong* and have been for **over three years**: https://github.com/google/material-design-icons/issues/1092.  Thanks, Google!

**Comment by dhdaines** [comment]
The workable solution for self-hosting appears to be https://github.com/marella/material-icons these days.

**Comment by JoshuaWi** [comment]
Hi all, I came here because I noticed my (svg) icon has a slight delay in displaying. I have a requirement to generate an image from the node and currently I am kicking this off in the ngAfterViewInit.. unfortunately it seems like the icon has not loaded in yet at this point. Does anyone have a way to hook into when the icon has actually finished loading? is there some event trigger we can subscribe to?
